neqo-crypto/src/constants.rs

```diff
@@ -6,21 +6,60 @@
 
 #![allow(dead_code)]
 
+use enum_map::Enum;
+
 use crate::ssl;
 
 // Ideally all of these would be enums, but size matters and we need to allow
 // for values outside of those that are defined here.
 
 pub type Alert = u8;
 
-pub type Epoch = u16;
-// TLS doesn't really have an "initial" concept that maps to QUIC so directly,
-// but this should be clear enough.
-pub const TLS_EPOCH_INITIAL: Epoch = 0_u16;
-pub const TLS_EPOCH_ZERO_RTT: Epoch = 1_u16;
-pub const TLS_EPOCH_HANDSHAKE: Epoch = 2_u16;
-// Also, we don't use TLS epochs > 3.
-pub const TLS_EPOCH_APPLICATION_DATA: Epoch = 3_u16;
+#[derive(Default, Debug, Enum)]
+pub enum Epoch {
+    // TLS doesn't really have an "initial" concept that maps to QUIC so directly,
+    // but this should be clear enough.
+    #[default]
+    Initial = 0,
+    ZeroRtt,
+    Handshake,
+    ApplicationData,
+    // Also, we don't use TLS epochs > 3.
+}
+
+impl From<u16> for Epoch {
+    fn from(e: u16) -> Self {
+        match e {
+            0 => Self::Initial,
+            1 => Self::ZeroRtt,
+            2 => Self::Handshake,
+            3 => Self::ApplicationData,
+            _ => unreachable!(),
+        }
+    }
+}
```

Comment:
Instead of implementing `From<u16>` one could as well implement `TryFrom<u16>`. That would remove the `unreachable!` here, and would instead force the caller to handle the error case.

I would assume the caller knows better whether the input is sanitized, i.e. whether panicing is safe.

Thoughts?

Reply:
Made the change. Also improved the other error handling in those functions a bit.